### PR TITLE
nixos/fresh-rss: init module

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/issues/77246

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @schmittlauch

### CODE REVIEW

File: nixos/modules/services/web-apps/fresh-rss.nix [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            A file containing the password corresponding to
+            <option>database.user</option>.
+          '';
+        };
+
+        createLocally = mkOption {
+          type = types.bool;
+          default = true;
+          description = ''
+            Create the database and database user locally.
```
Comment (schmittlauch) [inline nixos/modules/services/web-apps/fresh-rss.nix]: To me this description doesn't make completely clear what this option is about. Together with the assertion below I assume that this is used to create a mysql db on the same machine with socket authentication?

File: nixos/modules/services/web-apps/fresh-rss.nix [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    services.phpfpm.pools.fresh-rss = {
+      inherit user group;
+      settings = {
+        "listen.owner" = config.services.httpd.user;
+        "listen.group" = config.services.httpd.group;
+      } // cfg.poolConfig;
+    };
+
+    services.httpd = {
+      enable = true;
```
Comment (schmittlauch) [inline nixos/modules/services/web-apps/fresh-rss.nix]: So it looks like you always activate Apache httpd and use fresh-rss behind it. While this is definitely convenient for quickly getting a full RSS reader service running, it takes away the flexibility.
Other modules like gitea or tt-rss either don't care about the HTTP webserver in front of them at all or do supply a virtual host config for a web server of their choice (in the case of tt-rss it's nginx), but don't enable it by default.
This makes it possible for users to supply an alternative web server frontend.

File: pkgs/servers/web-apps/fresh-rss/default.nix [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  src = fetchFromGitHub {
+    owner = "FreshRSS";
+    repo = "FreshRSS";
+    rev = version;
+    sha256 = "0xnflvqhifk0icqqclv4qbs43x77syi3szha5mzqxf2fxyzrs358";
+  };
+
+  phpConfig = writeText "constants.local.php" ''
+  <?php
+    safe_define('DATA_PATH', '/var/lib/fresh-rss');
```
Comment (schmittlauch) [inline pkgs/servers/web-apps/fresh-rss/default.nix]: It'd be nice to tie this together with the `stateDir` variable defined in the module. Or is there a need to change the directories independently from each other in the package and module?

**Review by schmittlauch — COMMENTED** [review]
@aanderse First of all thank you for taking care of this so quickly.
These are my remarks after reading through your code *without* trying out the module so far.
Please also note that I'm relatively new to writing NixOS modules, so take my comments with a grain of salt.

File: pkgs/servers/web-apps/fresh-rss/default.nix (hunk 3 above)
Comment (Mic92) [reply pkgs/servers/web-apps/fresh-rss/default.nix]: It is not really configurable in the module either so this should be fine.

**Comment by aanderse** [comment]
@schmittlauch people who aren't as familiar with a technology often times provide the most valuable feedback as they have fresh eyes and are able to see shortcomings others can't :smile:

All feedback welcome. Thanks.

Note this is still a work in progress, though. I've done very minimal testing and I'm not convinced this works reliably yet.

More to come.

File: pkgs/servers/web-apps/fresh-rss/default.nix (hunk 3 above)
Comment (aanderse) [reply pkgs/servers/web-apps/fresh-rss/default.nix]: My intention is to leave this hardcoded. Whether a module has a configurable state directory or not is left up to the discretion of the module author. It is more work to make the state directory configurable, so I went with not.

Comment (schmittlauch) [reply pkgs/servers/web-apps/fresh-rss/default.nix]: It's just alway a potential source of issues if you define the same constant at multiple places and then forget to change both at some point.
But yeah, this is just a minor point.

File: nixos/modules/services/web-apps/fresh-rss.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/services/web-apps/fresh-rss.nix]: Some modules only provide `nginx` configuration, some only provide `httpd. Some provide neither. I find this situation regrettable and hope for a day where sysadmins can easily choose and modules support either option.

File: nixos/modules/services/web-apps/fresh-rss.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/web-apps/fresh-rss.nix]: Thanks for this feedback. You are correct. Any suggestions?

Comment (schmittlauch) [reply nixos/modules/services/web-apps/fresh-rss.nix]:
```nix
localDB = mkOption {
  type = types.bool;
  default = true;
  description = ''
    Create and use a database on the same machine. This allows using authetication to the db without any credentials.
    This currently only applies if database type "mysql" is selected.
  '';
};
```

**Comment by makefu** [comment]
Hey! I'd love to see this module being included into nixpkgs. Right now we "only" have tt-rss which has a couple of shortcomings and freshrss has a lot of potential.
Cheers!

**Comment by aanderse** [comment]
@makefu any interest in picking this up? I don't use `fresh-rss` so this PR is lower on my priorty list.

**Comment by makefu** [comment]
@aanderse low prio is fine for me, i will use ttrss until the time comes :) but thanks for even starting the PR!

**Comment by Ambroisie** [comment]
I am also interested in this MR, can I know what's left to be done?

**Comment by aanderse** [comment]
I don't think many people use `httpd` so someone should probably port it to `nginx`. Aside from that... probably just testing?

I don't use this software and have very little interest in finishing this off anymore. Competing priorities, etc... Are you interested in taking this over?

**Comment by Ambroisie** [comment]
@aanderse I'm quite new to Nix, and have never packaged anything for `nixpkgs`. 

I could probably port it to `nginx`, but I'm trying to understand why the service itself should be taking care of this, instead of being left to the user ?

I assume testing is mostly making sure the derivation builds, and probably also running correctly.

**Comment by aanderse** [comment]
The package is probably good to go. The module does a fair bit of work for the user. We could just ship a package without a module...

**Comment by aanderse** [comment]
If anyone wants to finish this off please feel free to take over.

**Pull request closed without merging**